ui.py

```diff
@@ -1,75 +1,102 @@
+        return get_copilot_view_setting(self.view, "completions", [])
 
     @property
-    def text(self) -> str:
-        return get_copilot_view_setting(self.view, "text") or ""
+    def completion_index(self) -> int:
+        """The index of the current chosen completion."""
+        return get_copilot_view_setting(self.view, "completion_index", 0)
 
     @property
-    def display_text(self) -> str:
-        return get_copilot_view_setting(self.view, "display_text") or ""
-
-    @property
-    def uuid(self) -> str:
-        return get_copilot_view_setting(self.view, "uuid") or ""
-
-    def get_display_text(self, region: Tuple[int, int], raw_display_text: str) -> str:
-        if "\n" in raw_display_text:
-            return raw_display_text
-
-        if self.view.classify(region[1]) & sublime.CLASS_LINE_END:
-            return raw_display_text
-
-        current_line = self.view.line(region[1])
-        following_text = self.view.substr(sublime.Region(region[0], current_line.end())).strip()
-        index = raw_display_text.find(following_text)
+    def current_completion(self) -> Optional[CopilotPayloadCompletion]:
+        """The current chosen `completion`."""
+        return self.completions[self.completion_index] if self.completions else None
+
+    def show_previous_completion(self) -> None:
+        """Show the previous completion."""
+        self.show(
+            completion_index=self.completion_index - 1,
+            do_clamp=not self.view.settings().get("auto_complete_cycle", False),
+        )
 
-        return raw_display_text[:index] if following_text and index != -1 else raw_display_text
+    def show_next_completion(self) -> None:
+        """Show the next completion."""
+        self.show(
+            completion_index=self.completion_index + 1,
+            do_clamp=not self.view.settings().get("auto_complete_cycle", False),
+        )
 
     def hide(self) -> None:
-        set_copilot_view_setting(self.view, "is_visible", False)
-
-        PopupCompletion.hide(self.view)
-
-    def show(self, region: Tuple[int, int], completions: List[CopilotPayloadCompletion], cycle: int = 0) -> None:
-        if not completions:
+        """Hide Copilot's completion popup."""
+        # prevent from hiding other plugin's popup
+        if self.is_visible:
+            _PopupCompletion.hide(self.view)
+
+    def show(
+        self,
+        completions: Optional[List[CopilotPayloadCompletion]] = None,
+        completion_index: Optional[int] = None,
+        do_clamp: bool = True,
+    ) -> None:
+        """Show Copilot's completion popup."""
+        # update completions
+        if completions is not None:
+            set_copilot_view_setting(self.view, "completions", completions)
+        # update completion index
+        if completion_index is not None:
+            set_copilot_view_setting(self.view, "completion_index", completion_index)
+        self._tidy_completion_index(do_clamp=do_clamp)
```

Comment:
Maybe just `(do_clamp)`?

Reply:
Done in [0fe023f](https://github.com/TheSecEng/LSP-copilot/pull/13/commits/0fe023fa1091bb206751cc838017dbe61d3373bd)